File: frontend/src/api/feedback.ts
import type { UserFeedbackCreate, UserFeedbackItem } from '../types/api';
import { requestAdminJson } from './adminClient';
import { requestJson } from './client';

interface BackendUserFeedbackItem {
  id: number;
  rating: number;
  content: string;
  created_at: string;
}

interface FeedbackListResponse {
  items: BackendUserFeedbackItem[];
  total: number;
}

interface FeedbackCreatePayload {
  rating: number;
  content: string;
  page_path: string;
}

export async function submitUserFeedback(payload: UserFeedbackCreate): Promise<UserFeedbackItem> {
  const response = await requestJson<BackendUserFeedbackItem, FeedbackCreatePayload>('/api/feedback', {
    method: 'POST',
    body: {
      rating: payload.rating,
      content: payload.content,
      page_path: payload.pagePath ?? '',
    },
  });
  return normalizeFeedback(response);
}

export async function getUserFeedback(): Promise<UserFeedbackItem[]> {
  const response = await requestAdminJson<FeedbackListResponse>('/api/feedback');
  return response.items.map(normalizeFeedback);
}

function normalizeFeedback(item: BackendUserFeedbackItem): UserFeedbackItem {
  return {
    id: String(item.id),
    rating: item.rating,
    content: item.content,
    createdAt: item.created_at,
  };
}

File: backend/app/api/feedback.py
from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.deps import get_db, require_admin
from app.models import UserFeedback
from app.schemas import UserFeedbackCreate

router = APIRouter(prefix="/api/feedback", tags=["feedback"])


@router.post("")
def create_feedback(payload: UserFeedbackCreate, db: Session = Depends(get_db)) -> dict:
    feedback = UserFeedback(
        rating=payload.rating,
        content=payload.content.strip(),
        page_path=payload.page_path.strip(),
    )
    db.add(feedback)
    db.commit()
    db.refresh(feedback)
    return {
        "id": feedback.id,
        "rating": feedback.rating,
        "content": feedback.content,
        "created_at": feedback.created_at.isoformat(),
    }


@router.get("", dependencies=[Depends(require_admin)])
def list_feedback(db: Session = Depends(get_db)) -> dict:
    items = db.scalars(
        select(UserFeedback).order_by(UserFeedback.created_at.desc(), UserFeedback.id.desc()),
    ).all()
    return {
        "items": [
            {
                "id": item.id,
                "rating": item.rating,
                "content": item.content,
                "created_at": item.created_at.isoformat(),
            }
            for item in items
        ],
        "total": len(items),
    }
